File: 2020/python/day6.py
from functools import reduce

with open("../input/input6.txt", "r") as fd:    
    groups = reduce(lambda a, x: (a[:-1] + [a[-1]+[x]]) if x else (a + [[]]), [x.strip() for x in fd], [[]])
    print("A:", sum([len(set(''.join(group))) for group in groups]))
    print("B:", sum([len(reduce(lambda l1, l2: set(l1).intersection(l2), group)) for group in groups]))

File: 2020/haskell/day6.hs
import qualified Data.Set as Set

group :: [String] -> [String] -> [[String]]
group [] a = [a]
group (x:xs) a = case length x of
        0 -> a : group xs []
        _ -> group xs (a ++ [x])

main = do
    c <- readFile "../input/input6.txt"
    let groups = group (lines c) []
    let sets = map (map Set.fromList) groups
    let a = sum $ map (Set.size . Set.unions) sets
    putStrLn ("A: " ++ show a)
    let b = sum $ map (Set.size . foldl1 Set.intersection) sets
    putStrLn ("B: " ++ show b)
